docs/enhancements/automatic-agent-classification-labels.md

```diff
@@ -0,0 +1,101 @@
+---
+title: automatic-agent-classification-labels
+authors:
+  - "@avishay"
+creation-date: 2022-02-27
+last-updated: 2022-02-27
+---
+
+# Automatic Agent Classification Labels
+
+## Summary
+
+When using late binding, users select which Agents should be part of a cluster based on their inventory.  The selection can be manual or by using a label selector.
+Two kinds of labels are helpful:
+* Inventory labels: Simply represents the hardware, such as architecture, virtualization support, etc.  The agent controller already adds these labels.
+* User-defined labels: These are labels that a user would add to characterize their hardware.  Examples include “t-shirt sizes” or hosts suitable for hosting storage or machine learning workloads.  There is currently no automation for adding these labels.
+
+This proposal is to automate user-defined labels.  Users would define the desired labels and conditions, and the system would automate the process.
+
+## Motivation
+
+Having users define labels manually is time-consuming and error-prone.  Defining conditions is more reliable and scalable for users.
+
+### Goals
+
+- Provide a flexible and easy-to-use API for defining desired labels and conditions.
+- Extend the Infrastructure Operator to maintain the desired labels based on the specified conditions with minimal impact on scale.
+
+### Non-Goals
+
+- Ship pre-defined labels and conditions.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a Cluster Creator, I want the system to automatically label Agents based on specific conditions so that I can more easily select Agents for cluster membership.	
+
+## Design Details [optional]
+
+We will create a new AgentClassification namespace-scoped CRD that defines the desired label key-value and a list of expressions to be evaluated according to the Agent’s inventory.  Each expression is generic such that it can refer to any properties in the inventory.
+
+### AgentClassification CRD Definition
```

Comment:
will it have a status indicating the validity of the expressions?

Reply:
Yes, forgot to add, thanks

Reply:
Added, please take a look

Reply:
Is there a condition that will validate the expressions? the status just show matched count. 
How will the user debug it?

Reply:
Added Conditions in the Status and defined QueryValidCondition const